### ISSUE TITLE
STOP TRAINING BASED ON VALIDATION ACCURACY OR EXTRACT MODEL TRAINED IN GIVEN EPOCH

### ISSUE THREAD

**Issue by msamwelmollel** [amaiya/ktrain#165]
Hello Amaya 

I am trying to use Ktrain to develop the NLP for sentimental analysis. However, my data has a high-class imbalance. I would like to know if is possible during training to extract only given epoch which has high validation accuracy to avoid overwriting. Let say I set the number of epoch to 10, and during training epoch 5 gives great validation accuracy and high comparative value for training accuracy. I want to use this model to predict the test set? How can I code the learner to give me this model trained in the 5th epoch

**Comment by amaiya** [comment]
The way you do this is similar to how you would normally do this in `tf.Keras`.  One way is to use the `checkpoint_folder` argument (wrapper to `ModelCheckpoint` callback) in any `*fit*` method to save weights of model for each epoch. Then, load the weights from whichever epoch you choose using `learner.model.load_weights`.  A second way is to use the `early_stopping` argument (wrapper to `EarlyStopping` callback) along with `monitor="val_accuracy"` which will load the weights from epoch with highest validation accuracy at end of training.

See the **Built-In-Callbacks** section in [this tutorial notebook[(https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-A1-additional-tricks.ipynb).

**Comment by amaiya** [comment]
Also, there is a `class_weight` argument in `*fit*` methods that can be used to addressed imbalanced datasets.   See [here](https://datascience.stackexchange.com/questions/13490/how-to-set-class-weights-for-imbalanced-classes-in-keras) for more info.
